src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,231 @@
+    {
+        private IDiagnosticAnalyzerService _diagnosticAnalyzerService;
+
+        public AbstractAddMissingImportsFeatureService(IDiagnosticAnalyzerService diagnosticAnalyzerService)
+        {
+            _diagnosticAnalyzerService = diagnosticAnalyzerService;
+        }
+
+        protected abstract ImmutableArray<string> FixableDiagnosticIds { get; }
+
+        public async Task<Solution> AddMissingImportsAsync(Solution solution, CancellationToken cancellationToken)
+        {
+            var newSolution = solution;
+
+            foreach (var projectId in solution.ProjectIds)
+            {
+                var project = newSolution.GetProject(projectId);
+                var newProject = await AddMissingImportsAsync(project, cancellationToken).ConfigureAwait(false);
+                newSolution = newProject.Solution;
+            }
+
+            return newSolution;
+        }
+
+        public async Task<Project> AddMissingImportsAsync(Project project, CancellationToken cancellationToken)
+        {
+            var newProject = project;
+
+            foreach (var documentId in project.DocumentIds)
+            {
+                var document = newProject.GetDocument(documentId);
+                newProject = await AddMissingImportsAsync(document, cancellationToken).ConfigureAwait(false);
+            }
+
+            return newProject;
+        }
+
+        public async Task<Project> AddMissingImportsAsync(Document document, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, cancellationToken).ConfigureAwait(false);
+            return await FixDocument(document, diagnostics, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<Project> AddMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            return await FixDocument(document, diagnostics, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<bool> IsMissingImportsAsync(Document document, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, cancellationToken).ConfigureAwait(false);
+            return !diagnostics.IsEmpty;
+        }
+
+        public async Task<bool> IsMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            return !diagnostics.IsEmpty;
+        }
+
+        private Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync(Document document, CancellationToken cancellationToken)
+        {
+            if (!document.TryGetText(out var text))
```

Comment:
what guarantees text being there? usually, with tree transformation, text will not be there. if only reason you do this is to get full span, you should try both tree and text. and if none of them exist, then GetSyntaxTreeAsync will be cheaper one to use. since it will always in memory once created rather than moving back and forth to same memory. or with tree transformation, might likely in memory.

by the way, with the way you implemented, AddMissing... Async methods above, I dont think most of time, text will exist in document. 

if you want to keep this way, I think this method should changed to TryGetDiagnosticAsync to indicate it failed rather than return default Empty.